ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java

```diff
@@ -2450,6 +2435,9 @@ public void testCleanerForTxnToWriteId() throws Exception {
     // Keep an open txn which refers to the aborted txn.
     Context ctx = new Context(hiveConf);
     HiveTxnManager txnMgr = TxnManagerFactory.getTxnManagerFactory().getTxnManager(hiveConf);
+    // Txn is not considered committed or aborted until TXN_OPENTXN_TIMEOUT expires
+    // See MinOpenTxnIdWaterMarkFunction, OpenTxnTimeoutLowBoundaryTxnIdHandler
+    waitUntilAllTxnFinished();
```

Comment:
why do we need to wait for all txn to commit here?

Reply:
The next command is `txnMgr.openTxn(ctx, "u1");`. During txn opening a new record is inserted into `MIN_HISTORY_LEVEL`
https://github.com/apache/hive/blob/cf83d4751271622a9a700c6f2330dbfff38801d2/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/jdbc/functions/OpenTxnsFunction.java#L150
It has an impact cleaning records from `TXN_TO_WRITE_ID`

The column value `MHL_MIN_OPEN_TXNID` is coming from 
https://github.com/apache/hive/blob/cf83d4751271622a9a700c6f2330dbfff38801d2/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/jdbc/functions/OpenTxnsFunction.java#L101C26-L101C55
It is calculated by taking the smaller value from `minOpenTxn` and `lowWaterMark`. `minOpenTxn` is the last open txn id which his null so we go with `lowWaterMark`.
`lowWaterMark` is queried by 
https://github.com/apache/hive/blob/cf83d4751271622a9a700c6f2330dbfff38801d2/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/jdbc/queries/OpenTxnTimeoutLowBoundaryTxnIdHandler.java#L42-L43

Maybe my test computer is running the queries with TEZ faster than MR and `openTxnTimeOutMillis` is not expired.
When I added `Thread.sleep(openTxnTimeOutMillis)` the test passed so I assume this is a timing issue.

Reply:
this part is a simulation of the following sequence of events:
````
commit-abort-open-abort-commit
````
I am not sure why we need to wait here since commit & abort are finite states

Reply:
You can repro the issue on the master branch using MR. Just set
`metastore.txn.opentxn.timeout` high enough. The default is 1sec and the insert statements in the test run  within 1 sec with Tez but more than 1 sec with MR on my test PC.

Reply:
I'll check

Reply:
I'm sorry, it might take a bit more time, but I don't want to hold it as a blocker. So, if everything else is resolved, let's merge it and I'll come back with the findings later if any.

Reply:
@kasakrisz,  i've checked out your branch and removed `waitUntilAllTxnFinished`, but still can't repro, even tried to increase `metastore.txn.opentxn.timeout` to 2 sec

Reply:
note, that `runCleaner` has the same wait logic before execution